include/envoy/buffer/buffer.h

```diff
@@ -125,13 +125,17 @@ class Instance {
 
   /**
    * Fetch the raw buffer slices. This routine is optimized for performance.
-   * @param out supplies an array of RawSlice objects to fill.
-   * @param out_size supplies the size of out.
-   * @return the actual number of slices needed, which may be greater than out_size. Passing
-   *         nullptr for out and 0 for out_size will just return the size of the array needed
-   *         to capture all of the slice data.
+   * @param out supplies an array of RawSlice objects to fill. Must not be nullptr.
+   * @param out_size supplies the size of out. Must be greater than 0.
+   * @return the number of slices written to |out|.
    */
-  virtual uint64_t getRawSlices(RawSlice* out, uint64_t out_size) const PURE;
+  virtual uint64_t getAtMostNRawSlices(RawSlice* out, uint64_t out_size) const PURE;
```

Comment:
QQ: WDYT about having a single function that takes either an absl InlinedArray or FixedArray as a reference param and an optional at_most param? Then I think we could fold both functions into 1 and not heap allocate anything in the common case?

Reply:
absl::FixedArray takes a size as a constructor argument and can't be changed beyond that, so it doesn't seem suitable.
absl::InlinedVector could work.  We would need to pick a value for the inlined size parameter.

Most of the uses of the new getRawSlices that returns a vector are in for-range loops, so a future optimization would be to instead return absl::iterator_range or some other way to get a const iterable for non-empty slices or all slices.

Reply:
Right makes sense that FixedArray won't work. I think InlinedArray with a pre-defined size of something like 16 (?) would not heap allocate in the vast majority of cases. WDYT?

Reply:
It prevents the idiomatic, common use of:
for (const RawSlice& slice : data.getRawSlices()) {

Equivalent version:
RawSliceVector slices;
data.getRawSlices(&slices);
for (const RawSlice& slice : slices) {

using RawSliceVector = absl::InlinedArray<RawSlices, 16>;

Reply:
what about a functional approach to prevent allocations?
add 
```
void iterateRawSlices(const std::function<void(RawSlice&)>);
```
and iteration becomes:
```
  std::string body;
 data.iterateRawSlices([&](RawSlice& slice){
    body += string_view(static_cast<const char*>(slice.mem_), slice.len_);
});
```

Reply:
yuval: You'ld need to allocate the function object at the call site in cases where the function has a non-null capture.

If we want to avoid allocation, we should consider going straight to the iterator range approach.

Reply:
> If we want to avoid allocation, we should consider going straight to the iterator range approach.

+1 if you are up for it. If we are going to refactor might as well go all the way? Maybe timebox it in case you run into anything super difficult and we can revisit?

Reply:
AFAIK these can be on the stack... no?
i.e. IIUC the lambda in my example above will most likely have no heap allocations assuming the size of the captured variables are below a certain (compiler defined) threshold?

Reply:
mattklein123: I'll look into the iterator approach
yuval-k: I'm no expert in how each compiler out there handles lambdas and std::functions.  My understanding is that std::function must be copyable.  It's possible that it manages to do things in the stack in this case.

Reply:
re: @yuval-k comment, AFAIK every modern compiler implements SVO (small value optimization) for lambas which would stack allocate the example you have above. You could confirm using godbolt or similar.

Reply:
Thinking about this more: an iterator approach is hard because we end up going through an interface.  All buffer implementations effectively need to share an iterator implementation.

I don't see an a simple idiomatic solution that supports range iteration better than the vector return approach.  `void iterateRawSlices(const std::function<void(RawSlice&)>&)` gets close.  `void iterateRawSlices(const std::function<bool(RawSlice&)>&)` would allow the caller to break out of the iteration if an error or termination condition is reached prior to consuming the full input.

Your call: absl::InlineVector, std::vector return or either iterateRawSlices approach.

Reply:
I would skip std::vector for perf reasons. Between the other 2 I'm fine either way. If you prefer a no prior declaration approach, I think the iterate solution will avoid any heap allocation in the common case.

Reply:
QQ: do we still need to keep the libevent implementation of buffer? If not maybe we can make the 'new' buffer implementation just be a concrete type. Would that simplify things?

Reply:
There are only 2 buffer implementations: the real one and one in some fuzzer test.  libevent is gone.

I tried the iterateRawSlices approach and ran into some problems in the zlib compressor and grpc inspector.  I'm halfway through the switch to absl::InlinedVector, updated PR should go out a bit later today.

Reply:
Switched to InlineVector.
Also added another interface method to get number of non-empty slices for the benefit of QUIC code and tests.  Hopefully we can remove that later.

Do you feel strongly about merging getRawSlices and getAtMostNRawSlices by switching FixedArray cases to InlineVector and taking the optional at_most argument?

Reply:
> Do you feel strongly about merging getRawSlices and getAtMostNRawSlices by switching FixedArray cases to InlineVector and taking the optional at_most argument?

I don't feel strongly about it, though I do think it would clean up the interface and yield the same perf (as well as be a more memory safe interface). @jmarantz WDYT?

Reply:
WDYT of passing in a pre-reserved InlineVector, and the function will just fill in to v.capacity(), and not bother to pass a separate integer arg?

I still like calling it getAtMostNRawSlices, but you'd just be forcing consistency by expressing the N as the vector's capacity.

Reply:
New version out. Being able to use InlinedVector and call getRawSlices directly in for-range loops is great.

getAtMostNRawSlices and numSlicesComputedSlowly are also gone.